src/midi/midi_mt32.cpp

```diff
@@ -378,98 +367,40 @@ void MidiHandler_mt32::PlaySysex(uint8_t *sysex, size_t len)
 	service->playSysexAt(sysex, msg_len, GetMidiEventTimestamp());
 }
 
-int MidiHandler_mt32::ProcessingThread(MAYBE_UNUSED void *data)
+// Ensures the member 'buffer' still has unplayed frames, and returns the quantity
+uint16_t MidiHandler_mt32::PopBuffer()
```

Comment:
This function name is really confusing considering the comment before. When reading code I kept thinking that PopBuffer does what `pop` operation usually does: returns the value from buffer (not number of elements).

Why comment in this function refer to "next buffer"?

Also, I think member variable `played_buffers` is confusingly named (shouldn't it be `played_frames` or something?).

Reply:
Suggestion: maybe this method should named something like `GetBufferredFramesNumber`?

Reply:
Thanks @dreamer. I will work in re-naming these and adding comments.

> Also, I think member variable **played_buffers** is confusingly named (shouldn't it be **played_frames**?)

The ring operates on "full buffer" objects. It actually holds three distinct `buffer_t` objects (each of  which inside holds 2048 frames).

Playback (in the mixer callback) happens at the frame-level. We steadily chew through the requested numbers of frames (in the current popped buffer), until we've played all its frames, and then "pop" out the next buffer to give us more frames to play (if a buffer isn't available then we're block and wait for the ring to give us one).

The moment we pop it out a new buffer for playing, back-pressure is released in the ring and MT-32 is unblocked to start rendering a new "full buffer", which is added to the ring.

We need to keep track of our actual played position, so this is done with two counters:
 - total "full buffers" played
 - plus the current frame played in the current buffer
 - The two are combined: `played_buffers * FRAMES_PER_BUFFER + current_frame`, to give us the total  frames played, which MT-32 uses to compute a time offset.

But yeah - these aren't obvious enough and I'll work on making it clearer.

Reply:
@dreamer , see the changes in: https://github.com/dosbox-staging/dosbox-staging/pull/863/commits/71d16408c848185d3d872ee0673997a477f3b6bf

Once we're done, I will squash this (and any additional fixes) down into the original commit.

Reply:
Ehh, I read these new changes and I still don't fully understand this algorithm… Why do we need to use a circular buffer to store 3 array buffers storing `int16_t` samples instead of simply storing the samples directly in the circular buffer? Isn't this an unnecessary complication? Am I missing something?

I guess maybe old code used 3 buffers, so mutex operations could be avoided in practice… but if we have fancy circular buffer that can be used from multiple threads at the same time, then having 3 buffers is pointless, I think?

Reply:
Yes, we could enqueue and dequeue individual samples directly, but then back pressure on the rendered would then paced per-sample: literally individually, and we don't want to be rendering individual samples because it's extremely inefficient.

Even operating on 1ms block-sizes (every 44 or 48 frames), is still extremely inefficient.

The old code also rendered in the same large chunk sizes, but didn't have the queue controlling the threads - so that's why the book keeping and thread control was so nasty.  

The win (in this bit of code) is the blocking queue being able to pace the threads, as opposed to using microscopic per-sample chunks (which I would do, if it were efficient... But like most things in audio they like to read or write to  cache-line-sized chunks of data). 

When it comes to the modem FIFO: hardware could read and write individual bytes simultaneously, but DOSBox doesn't because the exists FIFO s aren't thread safe. So that's another area where this queue can help and we will absolutely need to operate per-byte 👍

Reply:
We're eating through a full queue every 42 ms, at which point the MT32 is asked to start rendering another buffer-worth. 

It's not a lot of work for the queue or the threads.. which is a good thing to keep their involvement as minimal as possible given the MT32 is very CPU intensive (needing roughly 900 to 1000 MHz; it's too much to even run on the Mister FPGA and instead one full Pi3 core is essentially consumed).

To put it another way: if someone submitted a PR to render individual samples, my first comment would be to chunk it to 2 or 4 KiB :-)